resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterSchedulerBackend.scala

```diff
@@ -0,0 +1,427 @@
+        for (i <- 0 until math.min(
+          currentTotalExpectedExecutors - runningExecutorsToPods.size, podAllocationSize)) {
+          val executorId = EXECUTOR_ID_COUNTER.incrementAndGet().toString
+          val executorPod = executorPodFactory.createExecutorPod(
+            executorId,
+            applicationId(),
+            driverUrl,
+            conf.getExecutorEnv,
+            driverPod,
+            currentNodeToLocalTaskCount)
+          require(executorPod.getMetadata.getLabels.containsKey(SPARK_EXECUTOR_ID_LABEL),
+            s"Illegal internal state for pod with name ${executorPod.getMetadata.getName} - all" +
+              s" executor pods must contain the label $SPARK_EXECUTOR_ID_LABEL.")
+          val resolvedExecutorIdLabel = executorPod.getMetadata.getLabels.get(
+            SPARK_EXECUTOR_ID_LABEL)
+          require(resolvedExecutorIdLabel == executorId,
+            s"Illegal internal state for pod with name ${executorPod.getMetadata.getName} - all" +
+              s" executor pods must map the label with key ${SPARK_EXECUTOR_ID_LABEL} to the" +
+              s" executor's ID. This label mapped instead to: $resolvedExecutorIdLabel.")
+          executorsToAllocate(executorId) = executorPod
+          logInfo(
+            s"Requesting a new executor, total executors is now ${runningExecutorsToPods.size}")
+        }
+      }
+      val allocatedExecutors = executorsToAllocate.mapValues { pod =>
+        Utils.tryLog {
+          kubernetesClient.pods().create(pod)
+        }
+      }
+      RUNNING_EXECUTOR_PODS_LOCK.synchronized {
+        allocatedExecutors.map {
+          case (executorId, attemptedAllocatedExecutor) =>
+            attemptedAllocatedExecutor.map { successfullyAllocatedExecutor =>
+              runningExecutorsToPods.put(executorId, successfullyAllocatedExecutor)
+            }
+        }
+      }
+    }
+
+    def handleDisconnectedExecutors(): Unit = {
+      // For each disconnected executor, synchronize with the loss reasons that may have been found
+      // by the executor pod watcher. If the loss reason was discovered by the watcher,
+      // inform the parent class with removeExecutor.
+      disconnectedPodsByExecutorIdPendingRemoval.asScala.foreach {
+        case (executorId, executorPod) =>
+          val knownExitReason = Option(podsWithKnownExitReasons.remove(
+            executorPod.getMetadata.getName))
+          knownExitReason.fold {
+            removeExecutorOrIncrementLossReasonCheckCount(executorId)
+          } { executorExited =>
+            logWarning(s"Removing executor $executorId with loss reason " + executorExited.message)
+            removeExecutor(executorId, executorExited)
+            // We keep around executors that have exit conditions caused by the application. This
+            // allows them to be debugged later on. Otherwise, mark them as to be deleted from the
+            // the API server.
+            if (executorExited.exitCausedByApp) {
```

Comment:
It's not very clear to me when these executors will go away. Will the app eventually clean them up? Will the user have to manually do something?

Reply:
Executor pods are deleted when `stop` gets called.

Reply:
There are 2 different cases here. If the application succeeds, then as @liyinan926 said, we'll clean up all executors in `stop`. However, if for some reason, the application fails, then we will keep the failed executors and the failed driver around for debugging - this is the intended behavior here.

There are [owner references](https://kubernetes.io/docs/api-reference/v1.8/#ownerreference-v1-meta) between the driver pod and executors, such that the driver pod is the root of the ownership graph of all resources associated with a spark application. So, deleting the driver from the API automatically removes all executors. So, post-inspection, the expectation is that the user will delete the driver to simply clean up the application entirely. (We will include this in documentation)

In the future (not in the Spark 2.3 timeframe), the root of the owner reference graph may be a [CRD representing a Spark Application](https://github.com/apache-spark-on-k8s/spark/issues/111) which will provide more transparency to help manage resources associated with spark applications more intuitively.

Reply:
> we will keep the failed executors and the failed driver around for debugging

I understand the desire to make debugging easier, but isn't that bad, especially for long running applications? Won't you potentially end up with a bunch of containers holding on to CPU and memory, but no executor actually running on them?

Reply:
Actually no, the failed driver/executor containers have already finished running and won't take cpu/memory resources. The pod objects of failed driver/executors, however, are not deleted from the API server so users can use `kubectl logs <pod name>` or `kubectl describe pod <pod name>` to check what's going on with the failed pods. So they will take etcd storage but not cpu/memory resources.

Reply:
Ah, makes sense. It would be good to make the comment explain that.

Reply:
Done.